pallets/dapp-staking-v3/src/types.rs

```diff
@@ -170,108 +337,245 @@ where
-            unlocking: BoundedVec::<UnlockingChunk<Balance, BlockNumber>, UnlockingLen>::default(),
-            staked: (Balance::zero(), 0),
+            locked: SparseBoundedAmountEraVec(BoundedVec::<LockedChunk, LockedLen>::default()),
+            unlocking: BoundedVec::<UnlockingChunk<BlockNumber>, UnlockingLen>::default(),
+            staked: SparseBoundedAmountEraVec(BoundedVec::<StakeChunk, StakedLen>::default()),
+            staked_period: None,
         }
     }
 }
 
-impl<Balance, BlockNumber, LockedLen, UnlockingLen>
-    AccountLedger<Balance, BlockNumber, LockedLen, UnlockingLen>
+impl<BlockNumber, LockedLen, UnlockingLen, StakedLen>
+    AccountLedger<BlockNumber, LockedLen, UnlockingLen, StakedLen>
 where
-    Balance: AtLeast32BitUnsigned + MaxEncodedLen + Copy,
-    BlockNumber: AtLeast32BitUnsigned + MaxEncodedLen,
+    BlockNumber: AtLeast32BitUnsigned + MaxEncodedLen + Copy,
     LockedLen: Get<u32>,
     UnlockingLen: Get<u32>,
+    StakedLen: Get<u32>,
 {
     /// Empty if no locked/unlocking/staked info exists.
     pub fn is_empty(&self) -> bool {
-        self.locked.is_empty() && self.unlocking.is_empty() && self.staked.0.is_zero()
+        self.locked.0.is_empty() && self.unlocking.is_empty() && self.staked.0.is_empty()
     }
 
     /// Returns latest locked chunk if it exists, `None` otherwise
-    pub fn latest_locked_chunk(&self) -> Option<&LockedChunk<Balance>> {
-        self.locked.last()
+    pub fn latest_locked_chunk(&self) -> Option<&LockedChunk> {
+        self.locked.0.last()
     }
 
-    /// Returns locked amount.
-    /// If `zero`, means that associated account hasn't locked any funds.
-    pub fn locked_amount(&self) -> Balance {
+    /// Returns active locked amount.
+    /// If `zero`, means that associated account hasn't got any active locked funds.
+    pub fn active_locked_amount(&self) -> Balance {
         self.latest_locked_chunk()
             .map_or(Balance::zero(), |locked| locked.amount)
     }
 
+    /// Returns unlocking amount.
+    /// If `zero`, means that associated account hasn't got any unlocking chunks.
+    pub fn unlocking_amount(&self) -> Balance {
+        self.unlocking.iter().fold(Balance::zero(), |sum, chunk| {
+            sum.saturating_add(chunk.amount)
+        })
+    }
+
+    /// Total locked amount by the user.
+    /// Includes both active locked amount & unlocking amount.
+    pub fn total_locked_amount(&self) -> Balance {
+        self.active_locked_amount() + self.unlocking_amount()
+    }
+
     /// Returns latest era in which locked amount was updated or zero in case no lock amount exists
-    pub fn era(&self) -> EraNumber {
+    pub fn lock_era(&self) -> EraNumber {
         self.latest_locked_chunk()
             .map_or(EraNumber::zero(), |locked| locked.era)
     }
 
+    // TODO: can active_period be provided somehow different instead of using a parameter? It should be 'free' to read the data from storage since it will be whitelisted.
```

Comment:
Reading from storage would make it a bit difficult to do unit tests.

Reply:
I was thinking of adding another type with a `Get<...>` but that would make it even more complex 🙈 . Will keep it like this.